### ISSUE TITLE
Arms permitted when matching on uninhabited types

### ISSUE THREAD

**Issue by varkor** [rust-lang/rust#55123]
There seem to be two issues here:
```rust
pub enum Void {}

pub fn foo(x: Void) {
  match x {
    _ => {} // This arm shouldn't be permitted.
  };
  let _ = (); // This should be warned as unreachable, but isn't.
}
```
On the other hand, the following code *does* warn:
```rust
match () {
	() => {} // okay
	_ => {} // unreachable pattern
}
```

**Comment by varkor** [comment]
The second issue is a direct result of the first, where "number of arms" is treated as a heuristic for divergence, rather than the type of the expression (this could be an issue in the future with regards to an explicit `!` pattern):
https://github.com/rust-lang/rust/blob/3b88a54ebcf1d704fddd464e9d5225530cfb475b/src/librustc_typeck/check/_match.rs#L611-L615

**Comment by varkor** [comment]
This is due to: https://github.com/rust-lang/rust/blob/0982be77107f9af614adb609ee6eb1c5730d8eab/src/librustc_mir/hair/pattern/_match.rs#L374-L380 I think the solution here is to stabilise `exhaustive_patterns`.

**Comment by Nadrieril** [comment]
I don't think `exhaustive_patterns` would fix the second issue: it would detect the arm as unreachable, but afaik that bit of match checking does not count towards counting the match as divergent.

**Comment by Nadrieril** [comment]
The second issue got fixed in https://github.com/rust-lang/rust/pull/78995. I'm wondering about the first one: is matching on an empty enum very different from getting a value of an empty type as input? My hunch would be that if we lint any code after `match x {}` to be unreachable, why not lint the whole function body after `fn foo(x: Void)` as unreachable? Is there a particular reason that either is more natural/useful for an user?

```rust
pub enum Void {}

pub fn foo(x: Void) {
  let _ = (); // Frankly this also could be warned as unreachable.
  match x {};
  let _ = (); // This could be warned as unreachable, but isn't.
}
```

**Comment by varkor** [comment]
I agree that these are essentially the same issue. I actually submitted a PR a couple of years ago to address this, by making the uninhabitedness checking more eager (see #47291), but parts of it were split out to err on the side of being more conservative. I can't quite remember the status: probably in order to make these changes, there would need to be a discussion into how eagerly one wants to mark unreachable code based on uninhabited types. Probably this issue can be closed, and potentially a new one opened to explore reachability based on inhabitedness.

**Comment by camelid** [comment]
It looks like the `let _xyz` line isn't linted as unreachable because MIR for it is not generated (even in debug mode).

```rust
pub enum Void {}

pub fn foo(x: Void) {
  match x {
    _ => {} // This arm shouldn't be permitted.
  };
  let _xyz = (); // This should be warned as unreachable, but isn't.
}
```

**Comment by EFanZh** [comment]
Not sure if related, but I encountered two surprising error:

```rust
enum Empty {}

fn foo(x: &Empty) {
    match x {} // ERROR: missing match arm.
}
```

```rust
enum Empty1 {}

enum Empty2 {}

fn foo(x: (Empty1, Empty2)) {
    match x {} // ERROR: missing match arm.
}
```

I have to add a `_ => {}` arm for the code to compile.

**Comment by Nadrieril** [comment]
@EFanZh this is because in an unsafe block, you could match on a `&Empty` or a `(Empty1, u8)` without causing UB (if you're careful). Because of that we can't yet allow matching on those without any arms.
We want to make this possible though, but that'll take some work. More details here: https://github.com/rust-lang/rust/issues/51085
